src/eth1spec/trie.py

```diff
@@ -0,0 +1,279 @@
+`eth1spec.eth_types.Account` objects.
+"""
+
+from copy import copy
+from typing import Mapping, MutableMapping, Set, Union, cast
+
+from . import crypto, rlp
+from .base_types import U256, Bytes, Uint
+from .eth_types import Account, Receipt, Root
+
+debug = False
+verbose = False
+
+Node = Union[Account, Bytes, Receipt, Uint, U256]
+
+
+def nibble_list_to_compact(x: Bytes, terminal: bool) -> bytearray:
+    """
+    Compresses nibble-list into a standard byte array with a flag.
+
+    A nibble-list is a list of byte values no greater than `15`. The flag is
+    encoded in high nibble of the highest byte. The flag nibble can be broken
+    down into two two-bit flags.
+
+    Highest nibble:
+
+    ```
+    +---+---+----------+--------+
+    | _ | _ | terminal | parity |
+    +---+---+----------+--------+
+      3   2      1         0
+
+    ```
+
+    The lowest bit of the nibble encodes the parity of the length of the
+    remaining nibbles -- `0` when even and `1` when odd. The second lowest bit
+    encodes whether the key maps to a terminal node. The other two bits are not
+    used.
+
+    Parameters
+    ----------
+    x : `eth1spec.eth_types.Bytes`
+        Array of nibbles.
+    terminal : `bool`
+        Flag denoting if the key points to a terminal (leaf) node.
+
+    Returns
+    -------
+    compressed : `bytearray`
+        Compact byte array.
+    """
+    compact = bytearray()
+
+    if len(x) % 2 == 0:  # ie even length
+        compact.append(16 * (2 * terminal))
+        for i in range(0, len(x), 2):
+            compact.append(16 * x[i] + x[i + 1])
+    else:
+        compact.append(16 * ((2 * terminal) + 1) + x[0])
+        for i in range(1, len(x), 2):
+            compact.append(16 * x[i] + x[i + 1])
+
+    return compact
+
+
+def map_keys(
+    obj: Mapping[Bytes, Node], secured: bool = True
+) -> Mapping[Bytes, Node]:
+    """
+    Maps all compact keys to nibble-list format. Optionally hashes the keys.
+
+    Parameters
+    ----------
+    obj : `Dict[Bytes, T]`
+        Underlying trie key-value pairs.
+    secured : `bool`
+        Denotes whether the keys should be hashed. Defaults to `true`.
+
+    Returns
+    -------
+    out : `Mapping[Bytes, T]`
+        Object with keys mapped to nibble-byte form.
+    """
+    mapped: MutableMapping[Bytes, Node] = {}
+
+    # skip empty values, these are defined to be omitted from the trie
+    skip: Set[Bytes] = set()
+    for (k, v) in obj.items():
+        if v == b"":
+            skip.add(k)
+
+    for (preimage, value) in obj.items():
+        if preimage in skip:
+            continue
+
+        # "secure" tries hash keys once before construction
+        key = crypto.keccak256(preimage) if secured else preimage
+
+        nibble_list = bytearray(2 * len(key))
+        for i in range(2 * len(key)):
+            if i % 2 == 0:  # even
+                nibble_list[i] = key[i // 2] // 16
+            else:
+                nibble_list[i] = key[i // 2] % 16
```

Comment:
Just typing it out so that we can see whether it looks more readable using shifting and masking.  (not sure if this is actual equivalent logic)

```python
if is_even:
    nibbles_list[i] = key[i // 2] << 4
else:
    nibbles_list[i] = key[i // 2] & 0x0f
```

Code comments that just say "the high four bits" and "the low four bits" might be the most effective readability improvement here.